KIPs/kip-dreaft_sdk_common_architecture.md

```diff
@@ -0,0 +1,1623 @@
+| signMessage(messageHash: String): SignatureData | Signs the message. |
+| getPublicKey(compressed: Boolean): String | Returns public key string. |
+| getDerivedAddress(): String | Returns derived address from private key string. |
+
+#### AbstractKeyring
+
+`AbstractKeyring` is an abstract class that abstracts Keyring classes that store the account address and private key(s) to use. All Keyring classes are implemented by extending `AbstractKeyring`.
+
+##### Variable description
+
+| Variable | Description |
+| ----------- | ----------- |
+| address: String | The address of the account. |
+
+##### Method description
+
+| Method | Description |
+| ----------- | ----------- |
+| sign(txHash: String, chainId: int, role: int): List&#60;SignatureData&#62; | Signs the transaction using key(s) defined in role. `sign` is defined as an "abstract method" in AbstractKeyring, and must be implemented in all keyring classes that extend `AbstractKeyring`. |
+| sign(txHash: String, chainId: int, role: int, index: int): SignatureData | Signs the transaction using a key defined in role. `sign` is defined as an "abstract method" in AbstractKeyring, and must be implemented in all keyring classes that extend `AbstractKeyring`. |
+| signMessage(message: String, role: int): MessageSigned | Signs the message using key(s) defined in role. `signMessage` is defined as an "abstract method" in AbstractKeyring, and must be implemented in all keyring classes that extend `AbstractKeyring`. |
+| signMessage(message: String, role: int, index: int): MessageSigned | Signs the message using a key defined in role. `signMessage` is defined as an "abstract method" in AbstractKeyring, and must be implemented in all keyring classes that extend `AbstractKeyring`. |
+| encrypt(password: String, options: Object): Object | Encrypts a Keyring instance with keystore v4 format. `encrypt` is defined as an "abstract method" in AbstractKeyring, and must be implemented in all keyring classes that extend `AbstractKeyring`. |
+| copy(): AbstractKeyring | Returns a copied Keyring instance. `copy` is defined as an "abstract method" in AbstractKeyring, and must be implemented in all keyring classes that extend `AbstractKeyring`. |
+| sign(txHash: String, chainId: String, role: int): List&#60;SignatureData&#62; | Signs the transaction using key(s) defined in role. Converts the type of chainId inputted as String into int and calls the sign function implemented in keyring. |
+| sign(txHash: String, chainId: String, role: int, index: int): SignatureData | Signs the transaction using a key defined in role. Converts the type of chainId inputted as String into int and calls the sign function implemented in keyring. |
+| getKlaytnWalletKey(): String | Returns KlaytnWalletKey string. getKlaytnWalletKey is implemented to throw exception by default in `AbstractKeyring`. In the case of `SingleKeyring` that can use KlaytnWalletKey format, getKlaytnWalletKey function must be overridden. |
```

Comment:
```suggestion
| getKlaytnWalletKey(): String | Returns the KlaytnWalletKey string. It throws an exception by default in `AbstractKeyring`. |
```

Reply:
What about other functions? I think all the abstract functions should throw an exception.

Reply:
Yes, if abstract function is not implemented by sub classes, then exception should be thrown.
But getKlaytnWalletKey function is not defined as an abstract function.
getKlaytnWalletKey is implemented in AbstractKeyring class to throw exception as a default. (Because only SingleKeyring need to handle this)

Reply:
@jimni1222 
Okay. but the error message could be different.
"Not implemented" vs "not supported for this class"?

Reply:
We will unify the error message.

Defined as an abstract method, but if it is not implemented, "Not implemented" (caver-js only)

"Not supported for this class" when implemented to throw exception to abstract class as default because it is supported only in specific class

cc @sirano11